src/tools/rbd_nbd/win32/wnbd_ioctl.h

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2019 SUSE LLC
+ *
+ * Licensed under LGPL-2.1 (see LICENSE)
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <windows.h>
+#include <winioctl.h>
+
+/* WNBD Defines */
+#include "wnbd_shared.h"
+
+#define IOCTL_MINIPORT_PROCESS_SERVICE_IRP CTL_CODE(IOCTL_SCSI_BASE, 0x040e, METHOD_BUFFERED, FILE_READ_ACCESS | FILE_WRITE_ACCESS)
+
+DWORD
+WnbdUnmap(PCHAR instanceName);
+
+DWORD
+WnbdMap(PCHAR InstanceName,
+        PCHAR HostName,
+        PCHAR PortName,
+        PCHAR ExportName,
+        UINT64 DiskSize,
```

Comment:
The WNBD driver requires it and doesn't support live resizes either for the time being. We'll check if this limitation can be lifted.

Reply:
Is this a message being sent from `rbd(-nbd) map` to WNBD after loading the image and reading its metadata?

Reply:
Yep, we're opening the rbd image and doing a ``stat`` before that: https://github.com/petrutlucian94/ceph/blob/windows.12/src/tools/rbd_nbd/win32/rbd-nbd.cc#L667.

Reply:
FWIW, the NBD protocol currently doesn't support resizes after the connection has been established. Since WNBD uses the NBD protocol, that limitation can't be lifted without making an incompatible protocol change.

There is a proposal for a client side initiated resize request at https://github.com/NetworkBlockDevice/nbd/compare/extension-resize, but discussion on that proposal has stalled. If there is sufficient interest in that proposal from the Ceph project, I'm sure we can reopen discussion and finalize the protocol message...

Reply:
> FWIW, the NBD protocol currently doesn't support resizes after the connection has been established. Since WNBD uses the NBD protocol, that limitation can't be lifted without making an incompatible protocol change.
> 
> There is a proposal for a client side initiated resize request at https://github.com/NetworkBlockDevice/nbd/compare/extension-resize, but discussion on that proposal has stalled. If there is sufficient interest in that proposal from the Ceph project, I'm sure we can reopen discussion and finalize the protocol message...

Interesting. On Linux, rbd-nbd is currently using NBD_SET_SIZE to inform the driver that the image has been resized [1][2]. That probably means that NBD_SET_SIZE was supposed to be used only when establishing the connection, but not for resizes.

[1] https://github.com/ceph/ceph/blob/v16.0.0/src/tools/rbd_nbd/rbd-nbd.cc#L506-L517
[2] https://github.com/torvalds/linux/blob/v5.7/drivers/block/nbd.c#L1382

Reply:
Ah. No, `NBD_SET_SIZE` can be used to inform the kernel out of band that the device size is now different. Honestly, I wasn't even aware that rbd-nbd did that, nor that it was possible :-) what I'm talking about though is a proposed in-band protocol message that would work with more NBD clients than just the ceph NBD implementation.

I guess the downside of this method is that it's specific to a server which knows that the client is going to resize through out-of-band means.

Also, thinking about this, you guys probably want the server somehow to initiate a resize request, not the client. Ah well.

Reply:
Makes sense. The resize extension definitely looks interesting though, I'm wondering if it will become part of the standard. Not sure if Ceph would use this mechanism, but we can definitely consider it for WNBD as we intend to use it outside Ceph as well.

Issues may arise when shrinking images, especially when those aren't client initiated (e.g. corrupted filesystem). I'm not sure if this situation is handled by Ceph in any way, I'll need to look into it before enabling live resize support :).

Thanks a lot for taking the time to look over this work and thanks for the clarifications!